pkg/ebpf/c/bpf_helpers_custom.h

```diff
@@ -3,13 +3,45 @@
 
 #include "bpf_cross_compile.h"
 
-/* Macro to output debug logs to /sys/kernel/debug/tracing/trace_pipe
+// If we can, try to move the definition of the format string off
+// the stack and into the read-only section of the binary.
+#ifdef BPF_NO_GLOBAL_DATA
+#define BPF_PRINTK_FMT_MOD
+#else
+#define BPF_PRINTK_FMT_MOD static const
+#endif
+
+/*
+ * The existence of this tracepoint is used to detect if the bpf_trace_printk
+ * function adds a newline to the output or not (added in
+ * https://github.com/torvalds/linux/commit/ac5a72ea5c898, went upstream with 5.9)
+ *
+ * We define our own struct definition if our vmlinux.h is outdated.
+ * BPF CO-RE ignores the ___ and everything after it
+ */
+struct trace_event_raw_bpf_trace_printk___x {};
+
+#ifdef COMPILE_CORE
+#define BPF_PRINTK_ADDS_NEWLINE bpf_core_type_exists(struct trace_event_raw_bpf_trace_printk___x)
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 0)
+#define BPF_PRINTK_ADDS_NEWLINE 1
+#else
+#define BPF_PRINTK_ADDS_NEWLINE 0
+#endif
+
+/*
+ * Macro to output debug logs to /sys/kernel/debug/tracing/trace_pipe
  */
 #ifdef DEBUG
-#define log_debug(fmt, ...)                                        \
-    ({                                                             \
-        char ____fmt[] = fmt;                                      \
-        bpf_trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
+#define log_debug(fmt, ...)                                            \
+    ({                                                                 \
+        if (BPF_PRINTK_ADDS_NEWLINE) {                                 \
+            BPF_PRINTK_FMT_MOD char ____fmt[] = fmt;                   \
+            bpf_trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
+        } else {                                                       \
+            BPF_PRINTK_FMT_MOD char ____fmt[] = fmt "\n";              \
+            bpf_trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
+        }                                                              \
```

Comment:
my only concern that this adds additional instruction to ebpf code and might hitting the instruction limit in cases we use the `debug_log` on "hot-paths"

Reply:
not only instructions but duplicates the branches for the verifier (a problem in which we suffer heavily in http2 programs)

Reply:
The verifier should be doing dead code elimination here, but that is not supported on all kernels.

Reply:
Can't we do a macro for `\n` and then define it to nothing for when `BPF_PRINTK_ADDS_NEWLINE` is `0`?